File: server/src/Devilution.Server/Content/ContentManifest.cs
using Devilution.Server.Snapshots;

namespace Devilution.Server.Content;

/** One ordered external content pack and its validated tables. */
public sealed record ContentPack
{
    public ContentPack(string id, string version, IReadOnlyList<TsvTable> tables)
    {
        if (string.IsNullOrWhiteSpace(id) || string.IsNullOrWhiteSpace(version))
            throw new ArgumentException("Content pack IDs and versions are required.");
        Id = id;
        Version = version;
        Tables = tables ?? throw new ArgumentNullException(nameof(tables));
    }

    public string Id { get; }

    public string Version { get; }

    public IReadOnlyList<TsvTable> Tables { get; }
}

/** Deterministic identity for the ordered content set active on a server. */
public sealed class ContentManifest
{
    public ContentManifest(string id, string version, IEnumerable<ContentPack> packs)
    {
        if (string.IsNullOrWhiteSpace(id))
            throw new ArgumentException("A content manifest ID is required.", nameof(id));
        if (string.IsNullOrWhiteSpace(version))
            throw new ArgumentException("A content manifest version is required.", nameof(version));
        ArgumentNullException.ThrowIfNull(packs);

        Id = id;
        Version = version;
        Packs = packs.ToArray();
        if (Packs.Count == 0)
            throw new ArgumentException("A content manifest must contain at least one pack.", nameof(packs));
        if (Packs.Select(pack => pack.Id).Distinct(StringComparer.Ordinal).Count() != Packs.Count)
            throw new ArgumentException("Content pack IDs must be unique.", nameof(packs));
        Sha256 = ComputeHash();
    }

    public string Id { get; }

    public string Version { get; }

    public IReadOnlyList<ContentPack> Packs { get; }

    public string Sha256 { get; }

    private string ComputeHash()
    {
        var hasher = new CanonicalStateHasher();
        hasher.AppendString(Id);
        hasher.AppendString(Version);
        hasher.AppendUInt64((ulong)Packs.Count);
        foreach (var pack in Packs) {
            hasher.AppendString(pack.Id);
            hasher.AppendString(pack.Version);
            hasher.AppendUInt64((ulong)pack.Tables.Count);
            foreach (var table in pack.Tables) {
                hasher.AppendString(table.SourcePath);
                hasher.AppendUInt64((ulong)table.Columns.Count);
                foreach (var column in table.Columns)
                    hasher.AppendString(column);
                hasher.AppendUInt64((ulong)table.Rows.Count);
                foreach (var row in table.Rows) {
                    hasher.AppendUInt64((ulong)row.LineNumber);
                    foreach (var column in table.Columns)
                        hasher.AppendString(row[column]);
                }
            }
        }
        return hasher.HexDigest();
    }
}

File: Source/data/content_manifest.cpp
#include "data/content_manifest.hpp"

#include <array>
#include <cstdint>
#include <stdexcept>
#include <string_view>
#include <vector>

#include <picosha2.h>

namespace devilution {
namespace {

class CanonicalContentHasher {
public:
	void AppendString(std::string_view value)
	{
		AppendUint64(value.size());
		bytes_.insert(bytes_.end(), value.begin(), value.end());
	}

	void AppendUint64(uint64_t value)
	{
		for (size_t byte = 0; byte < sizeof(value); ++byte)
			bytes_.push_back(static_cast<uint8_t>(value >> (byte * 8)));
	}

	[[nodiscard]] std::string HexDigest() const
	{
		std::array<uint8_t, picosha2::k_digest_size> digest;
		picosha2::hash256(bytes_.begin(), bytes_.end(), digest.begin(), digest.end());
		return picosha2::bytes_to_hex_string(digest.begin(), digest.end());
	}

private:
	std::vector<uint8_t> bytes_;
};

void AppendTable(CanonicalContentHasher &hasher, const ContentManifestTable &table)
{
	hasher.AppendString(table.sourcePath);
	hasher.AppendUint64(table.columns.size());
	for (const std::string &column : table.columns)
		hasher.AppendString(column);
	hasher.AppendUint64(table.rows.size());
	for (const ContentManifestRow &row : table.rows) {
		if (row.cells.size() != table.columns.size())
			throw std::invalid_argument("Content manifest row cell count does not match the table column count.");
		hasher.AppendUint64(row.lineNumber);
		for (size_t column = 0; column < table.columns.size(); ++column)
			hasher.AppendString(row.cells[column]);
	}
}

} // namespace

std::string ComputeContentManifestSha256(const ContentManifest &manifest)
{
	CanonicalContentHasher hasher;
	hasher.AppendString(manifest.id);
	hasher.AppendString(manifest.version);
	hasher.AppendUint64(manifest.packs.size());
	for (const ContentManifestPack &pack : manifest.packs) {
		hasher.AppendString(pack.id);
		hasher.AppendString(pack.version);
		hasher.AppendUint64(pack.tables.size());
		for (const ContentManifestTable &table : pack.tables)
			AppendTable(hasher, table);
	}
	return hasher.HexDigest();
}

} // namespace devilution
